kuma/javascript/src/search-results-page.jsx

```diff
@@ -1,84 +1,206 @@
     locale: string,
-    query: string
+    query: string,
+    page: ?number
 };
 
-export type SearchResults = {
-    results: ?Array<{
+type SearchResults = {
+    count: number,
+    documents: ?Array<{
         slug: string,
         title: string,
-        summary: string,
-        score: number,
-        excerpts: Array<string>
+        locale: string,
+        excerpt: string
     }>,
+    start: number,
+    end: number,
+    filters: ?any, // TODO specify this maybe
+    locale: string,
+    next: ?string,
+    page: number,
+    pages: number,
+    previous: ?string,
+    query: string
+};
+export type SearchResultsResponse = {
+    results: ?SearchResults,
     error: ?any
 };
 
 type Props = {
     locale: string,
     query: string,
-    data: ?SearchResults
+    data: ?SearchResultsResponse
 };
 
+function makePaginationPageURL(uri) {
+    return uri ? `?${uri.split('?')[1]}` : '';
+}
+
 export default function SearchResultsPage({ locale, query, data }: Props) {
-    return (
-        <>
-            <Header searchQuery={query} />
-            <Titlebar title={`${gettext('Results')}: ${query}`} />
-            <div className="search-results">
-                {data &&
-                    data.results &&
-                    data.results.map(hit => {
-                        let path = `/${locale}/docs/${hit.slug}`;
-                        let url =
-                            window && window.origin
-                                ? `${window.origin}${path}`
-                                : path;
-                        return (
-                            <div className="result-container" key={hit.slug}>
-                                <div className="result">
-                                    <div>
-                                        <a className="result-title" href={path}>
-                                            {hit.title}
-                                        </a>
-                                    </div>
-                                    <div className="result-url">
-                                        <a href={path}>{url}</a>
-                                    </div>
-                                    <div className="result-summary">
-                                        {hit.summary}
-                                    </div>
-                                    {hit.excerpts.map((excerpt, i) => (
-                                        <div
-                                            className="result-excerpt"
-                                            key={i}
-                                            dangerouslySetInnerHTML={{
-                                                __html: excerpt
-                                            }}
-                                        />
-                                    ))}
-                                </div>
+    // Because it's clunky to do conditionals in JSX, let's create all the
+    // React nodes here first.
```

Comment:
I think it could be even less clunky, in terms of nesting and mismatch between declaration and insertion order. I'd personally go for more componentization.

Reply:
Not sure what you mean. 
*Instead* doing:
```jsx
    if (!query) {
        noQuery = (
            <div className="result-container">
                <p>
                    <i>{gettext('Nothing found if nothing searched.')}</i>
                </p>
            </div>
        );
    }
```
it could be done like this:
```jsx
    if (!query) {
        noQuery = <NoQuery/>
    }
```
But I'm not sure it's of any use. It would make the total length-of-lines-of-code smaller for this render function but that means you'd have to jump between here go and find that `NoQuery` component. Now, although big, it reads naturally and you don't have to jump around in the file.

I personally like making components out of things to break them out, but only when it really helps. Such as, if it's used in more than 1 place. 

There was a point when this render function had a bunch of things like this:

```jsx
{data && !query && <STUFF>}
{data && results && !results.error && (results.previous || results.next) && <STUFF>}
{data && results && !results.error && !results.length && <STUFF>}
...etc...
```

It just was getting uglier and uglier. The lack of expressing simple `else` statements was annoying. 

I'm not in love with the way it turned out but it's not too far from good. At least it *reads* OK.

Reply:
I do agree with your view around componentization and not having to jump around in files generally. But I think in this case one has to jump anyway, since declaration and insertion are at different points. And imo jumping around for understanding content, is better than jumping around for understanding control flow. So my idea of more readable would look more like this:
```jsx
export default function SearchResultsPage({ locale, query, data }: Props) {
  const results = data ? data.results : null;
  return (
    <>
      <Header searchQuery={query} />
      {query && <Titlebar title={`${gettext('Results')}: ${query}`} />}

      <div className="search-results">
        {data && (
          <>
            {results && (
              <>
                <ResultsMeta {...{ results, locale }} />

                <Results {...{ results, locale }} />

                {(results.previous || results.next) && (
                  <Pager results={results} />
                )}

                {results.count === 0 && <NoResults />}
              </>
            )}
          </>
        )}

        {data.error && <ResultError error={data.error} />}

        {!query && (
          <div className="result-container">
            <p>
              <i>{gettext('Nothing found if nothing searched.')}</i>
            </p>
          </div>
        )}
      </div>
    </>
  );
}
```

Anyway, this might be more subjective than I think and I totally agree about merging and moving along :)